compiler/qsc_formatter/src/formatter/tests.rs

```diff
@@ -686,6 +685,213 @@ fn formatting_corrects_indentation() {
+            (1)
+        ];
+            let y = 3;
+    "#},
+        &expect![[r#"
+            let x = [
+                // this is a comment
+                (1)
+            ];
+            let y = 3;
+    "#]],
+    );
+}
+
+#[test]
+fn brace_no_newlines() {
+    check(
+        indoc! {r#"
+        { Foo();
+            Bar(); Baz()
+        }
+    "#},
+        &expect![[r#"
+            { Foo(); Bar(); Baz() }
+        "#]],
+    );
+}
+
+#[test]
+fn brace_newlines() {
+    check(
+        indoc! {r#"
+        {
+            Foo(); Bar(); Baz() }
+    "#},
+        &expect![[r#"
+            {
+                Foo();
+                Bar();
+                Baz()
+            }
+        "#]],
+    );
+}
+
+#[test]
+fn parens_no_newlines() {
+    check(
+        indoc! {r#"
+        ( Foo(),
+            Bar(), Baz()
+        )
+    "#},
+        &expect![[r#"
+            (Foo(), Bar(), Baz())
+        "#]],
+    );
+}
+
+#[test]
+fn parens_newlines() {
+    check(
+        indoc! {r#"
+        (
+            Foo(), Bar(), Baz() )
+    "#},
+        &expect![[r#"
+            (
+                Foo(),
+                Bar(),
+                Baz()
+            )
+        "#]],
+    );
+}
+
+#[test]
+fn bracket_no_newlines() {
+    check(
+        indoc! {r#"
+        [ Foo(),
+            Bar(), Baz()
+        ]
+    "#},
+        &expect![[r#"
+            [Foo(), Bar(), Baz()]
+        "#]],
+    );
+}
+
+#[test]
+fn bracket_newlines() {
+    check(
+        indoc! {r#"
+        [
+            Foo(), Bar(), Baz() ]
+    "#},
+        &expect![[r#"
+            [
+                Foo(),
+                Bar(),
+                Baz()
+            ]
+        "#]],
+    );
+}
+
+#[test]
+fn semi_no_context_uses_newlines() {
+    check(
+        indoc! {r#"
+        Foo(); Bar(); Baz()
+    "#},
+        &expect![[r#"
+            Foo();
+            Bar();
+            Baz()
+        "#]],
+    );
+}
+
+#[test]
+fn comma_no_context_uses_space() {
+    check(
+        indoc! {r#"
+        Foo(),
+        Bar(),
+        Baz()
+    "#},
+        &expect![[r#"
+            Foo(), Bar(), Baz()
+        "#]],
+    );
+}
+
+#[test]
+fn type_param_lists_have_no_spaces_around_delims() {
+    check(
+        indoc! {r#"
+        {
+            operation Foo < 'A,
+            'B,   'C > (a : 'A, b : 'B, c : 'C) : Unit {}
+        }
+    "#},
+        &expect![[r#"
+            {
+                operation Foo<'A, 'B, 'C>(a : 'A, b : 'B, c : 'C) : Unit {}
+            }
+        "#]],
+    );
+}
+
+#[test]
+fn greater_than_and_less_than_bin_ops_have_spaces() {
+    check(indoc! {r#"x<y>z;"#}, &expect!["x < y > z;"])
+}
+
+#[test]
+fn delimiter_newlines_indentation() {
+    check(
+        r#"
+        let x = [ a,  b,  c ];
+        let y = [   a,
+            b,  c ];
+        let z = [
+
+
+            a,  b, c
+        ];
+"#,
+        &expect![[r#"
+            let x = [a, b, c];
+            let y = [a, b, c];
+            let z = [
+
+
```

Comment:
Not suggesting it as a change on this PR, but just curious: is it theoretically possible for us to have authoritative logic around collapsing multiple new lines into one? 

I have seen some formatters do it, and some don't. I can't even make up my mind about whether it's desirable.

Reply:
I think it should be possible, at least to stop more the 2 newlines from showing up anywhere. I should really put together a little todo task list for improvements like this to the formatter.

Reply:
Turns out I already had a todo-list style issue for the formatter: #1025. I've updated it a bit since I last looked at it.